dbms/src/Flash/Pipeline/Pipeline.cpp

```diff
@@ -32,6 +33,68 @@ FmtBuffer & addPrefix(FmtBuffer & buffer, size_t level)
 {
     return buffer.append(String(level, ' '));
 }
+
+void mapEvents(const Events & inputs, const Events & outputs)
+{
+    assert(!inputs.empty());
+    assert(!outputs.empty());
+    if (inputs.size() == outputs.size())
```

Comment:
A little strange condition branch here, seems not that straightforward, can we take fine grained info directly into consideration here?

Reply:
ok, refined as
```
struct PipelineEvents
{
    Events events;
    bool is_fine_grained;

    PipelineEvents(Events && events_, bool is_fine_grained_);

    void mapInputs(const PipelineEvents & inputs);
};
```